text/0000-mark-cf.md

```diff
@@ -0,0 +1,159 @@
+
+This is a tricky optimization and I really doubt if it is really worth.
+
+If a new `Lock` record is written to the write CF while the latest version is also a `Lock` or `Rollback`, instead of removing the previous version, just overwrite that record and add a `real_commit_ts` to the record. When checking write conflicts, we should parse the value and check the real commit TS because the timestamp encoded into the key may be not accurate.
+
+It may help reduce tombstones but breaks too many assumptions before.
+
+### Reading the mark CF
+
+Putting `Lock` and `Rollback` records in the mark CF does not introduce extra seek operations to the happy path of transactions. But it does when we talk about resolving locks.
+
+Now, the records that represent transaction status spread in both the write CF and the mark CF.
+
+In `CheckTxnStatus`, we need to read both CFs of the primary key for the status of the transaction.
+
+In `CheckSecondaryKeys`, we need to check both CFs of all the given secondary keys to know whether some keys are already committed or rolled back.
+
+And when prewrite raises an error or we are prewriting non-pessimistic keys in a retry, we also need the precise status of the key to guarantee idempotence. This also requires to read both the write and mark CFs.
+
+Luckily, all of these don't happen frequently in production. The extra cost is not a big issue.
+
+### No more overlapped rollback
+
+This is an optional change. Overlapped rollback describes the case when the `start_ts` of a `Rollback` record equals to the `commit_ts` of another write record, the `Rollback` record will become a special flag in that write record. We have implemented it, but it was proved to be a tricky solution as it caused incompatibilities with the compaction filter and CDC mainly because it overwrites an existing write record. 
+
+It is possible for us to get rid of it now. First, it's impossible for a normal write record to overwrite a protected `Rollback` in the write CF because we update the `max_ts` when writing a protected `Rollback`. So, we only need to consider overwriting by a new `Rollback`. Under such circumstances, we can just skip writing `Rollback` to the write CF. It does not affect conflict checking. And when we need the precise status of the transaction, the `Rollback` in the mark CF can also tell us.
+
+Because the keys in the mark CF are encoded with `start_ts` and `start_ts` is the unique identifier of transactions, it is impossible to have key conflicts in the mark CF.
+
+### Garbage collection
+
+The records in the mark CF don't need to exist after all keys in the transaction are totally resolved. The client resolves all the locks before a certain timestamp before updating this timestamp as the safe point.
+
+So, when TiKV is ready to do GC, all records in the mark CF whose `commit_ts` is less than the safe point can be deleted. It can be done in the compaction filter.
```

Comment:
Do we need to parse the value to get the real `commit_ts` for each key in the filter as the orders may be different?

Reply:
Yes. We need to parse the value to get the real `commit_ts`. The key itself cannot tell us.